src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,354 @@
+        return INITIAL_XRP;
+    // LCOV_EXCL_STOP
+
+    XRPAmount signerFees = signerCount * view.fees().base;
+
+    // LCOV_EXCL_START
+    if (signerFees > maxAmount - txnFees)
+        return INITIAL_XRP;
+    if (txnFees + signerFees > maxAmount - batchBase)
+        return INITIAL_XRP;
+    // LCOV_EXCL_STOP
+
+    // 10 drops per batch signature + sum of inner tx fees + batchBase
+    return signerFees + txnFees + batchBase;
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const parentBatchId = ctx.tx.getTransactionID();
+    auto const outerAccount = ctx.tx.getAccountID(sfAccount);
+    auto const flags = ctx.tx.getFlags();
+
+    if (flags & tfBatchMask)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    if (std::popcount(
+            flags &
+            (tfAllOrNothing | tfOnlyOne | tfUntilFailure | tfIndependent)) != 1)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "too many flags.";
+        return temINVALID_FLAG;
+    }
+
+    auto const& rawTxns = ctx.tx.getFieldArray(sfRawTransactions);
+    if (rawTxns.size() <= 1)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "txns array must have at least 2 entries.";
+        return temARRAY_EMPTY;
+    }
+
+    if (rawTxns.size() > maxBatchTxCount)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "txns array exceeds 8 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    // Validation Inner Batch Txns
+    std::unordered_set<AccountID> requiredSigners;
+    std::unordered_set<uint256> uniqueHashes;
+    std::unordered_map<AccountID, std::unordered_set<std::uint32_t>>
+        accountSequences;
+    std::unordered_map<AccountID, std::unordered_set<std::uint32_t>>
+        accountTickets;
+    for (STObject rb : rawTxns)
+    {
+        STTx const stx = STTx{std::move(rb)};
+        auto const hash = stx.getTransactionID();
+        if (!uniqueHashes.emplace(hash).second)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                                << "duplicate Txn found. "
+                                << "txID: " << hash;
+            return temINVALID_INNER_BATCH;
+        }
+
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                                << "batch cannot have an inner batch txn. "
+                                << "txID: " << hash;
+            return temINVALID_INNER_BATCH;
```

Comment:
Please add a test for this clause

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/743e1a1f7eddc740d4f7dca21027217f57073f1c